cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -400,6 +412,135 @@ async function activateOption(option, loadBalancerDNS, loadBalancerHostedZone) {
+    // NOTE: `Set` uses object reference to identify elements. That is, two options with the same value will not be
+    // identify as the same option. Therefore, we use map with object string as keys to identify object with actual
+    // values instead of object reference.
+    let optionsInUse = new Map();
+    let optionsPendingUnusedByService = new Map();
+
+    let certificates = await serviceCertificates();
+    for (const {DomainName: domainName, SubjectAlternativeNames: sans, DomainValidationOptions: validationOptions} of certificates) {
+        if (domainName === certificateDomain && setEqual(aliases.add(certificateDomain), new Set(sans))) {
+            // This is a certificate pending deletion. It is:
+            // 1. Copilot-tagged, 2. use the default certificate domain name, and 3. uses the SANs that need to be deleted.
+            for (let option of validationOptions) {
+                optionsPendingUnusedByService.set(JSON.stringify(option), option);
+            }
+            continue;
+        }
+        // Otherwise, this is a certificate that we do not want to touch; validation options of this certificate are deemed in use.
+        for (let option of validationOptions) {
+            optionsInUse.set(JSON.stringify(option), option);
+        }
+    }
+    for (const optionInUseKey of optionsInUse.keys()) {
+        optionsPendingUnusedByService.delete(optionInUseKey);
+    }
+    let options = new Set();
+    for (const optionPendingUnused of optionsPendingUnusedByService.values()) {
+        options.add(optionPendingUnused);
+    }
+    return options;
+}
+
+/**
+ * Retrieve validation options that will be unused by any service.
+ *
+ * @param {Set<String>} aliases
+ * @param {String} loadBalancerDNS
+ * @returns {Promise<Set<any>>}
+ */
+async function unusedValidationOptions(aliases, loadBalancerDNS) {
+    let optionsPendingUnused = await optionsUnusedByService(aliases);
+    let promises = [];
+    for (let option of optionsPendingUnused) {
+        let domainName = option["DomainName"];
+        let hostedZoneID, route53Client;
+        try {
+            ({hostedZoneID, route53Client} = await domainResources(domainName));
+        } catch (err) {
+            if (!err instanceof UnrecognizedDomainTypeError) {
+                throw err;
+            }
+            // This option has unrecognized pattern, we can't check if it is in use, so we skip deleting it.
+            optionsPendingUnused.delete(option);
+            continue;
+        }
+        const promise = route53Client.listResourceRecordSets({
+            HostedZoneId: hostedZoneID,
+            MaxItems: "1",
+            StartRecordName: domainName,
+        }).promise().then(({ ResourceRecordSets: recordSet }) => {
+            if (!targetRecordExists(domainName, recordSet)) {
+                return; // If there is no record using this domain, it is not in use.
+            }
+            let aliasTarget = recordSet[0].AliasTarget;
```

Comment:
Is there always only one record set?

Reply:
Yes because in the request we specified `MaxItems:1`. Therefore, they will return <= 1 recordset starting from `domainName`.